src/bobtail/dag.cpp

```diff
@@ -127,40 +189,44 @@ bool CDagForrest::Insert(const CSubBlock &sub_block)
             continue;
         }
         newNode->AddAncestor(ancestor);
-        merge_list.emplace(ancestor->subgraph_id);
+        merge_list.emplace(ancestor->dag_id);
         ancestor->AddDescendant(newNode);
     }
     int16_t new_id = -1;
     if (merge_list.size() > 1)
     {
-        new_id = MergeTrees(merge_list);
+        if (!MergeDags(merge_list, new_id))
+        {
+            return false;
+        }
     }
     else if (merge_list.size() == 1)
     {
         new_id = *(merge_list.begin());
     }
     else // if(merge_list.size() == 0)
     {
-        new_id = next_subgraph_id;
-        next_subgraph_id = next_subgraph_id + 1;
+        new_id = vdags.size();
+        vdags.emplace_back(new_id, newNode);
     }
-    newNode->subgraph_id = new_id;
+    newNode->dag_id = new_id;
+    vdags[new_id].Insert(newNode);
     // TODO : should insert to maintain temporal ordering not just emplace_back
-    _dag.emplace_back(newNode);
+    mapAllNodes.emplace(newNode->hash, newNode);
```

Comment:
Theoretically mapAllNodes could be a map of maps: `std::map<uint256, <uint16_t, CDagNode*>>`. I don't have strong feelings about this.

Reply:
its supposed to be an unordered map of all nodes in every dag. you can get the id of the dag from inside the node itself if needed

Reply:
I see. So would that be `bobtailDagSet.vdags[bobtailDagSet.Find(hash)->dag_id]`?

Reply:
to get the dag it was in, yes.